public/app/features/dashboard-scene/saving/DashboardPrompt.tsx

```diff
@@ -0,0 +1,188 @@
+    //       hideModal();
+    //       moveToBlockedLocationAfterReactStateUpdate(location);
+    //     },
+    //     onDiscard: () => {
+    //       dispatch(discardPanelChanges());
+    //       moveToBlockedLocationAfterReactStateUpdate(location);
+    //       hideModal();
+    //     },
+    //     onDismiss: hideModal,
+    //   });
+    //   return false;
+    // }
+
+    // Are we still on the same dashboard?
+    if (originalPath === location.pathname) {
+      return true;
+    }
+
+    if (ignoreChanges(dashboard, dashboard.getInitialSaveModel())) {
+      return true;
+    }
+
+    if (!dashboard.state.isDirty) {
+      return true;
+    }
+
+    showModal(UnsavedChangesModal, {
+      dashboard,
+      onSaveDashboardClick: () => {
+        hideModal();
+        dashboard.openSaveDrawer({
+          onSaveSuccess: () => {
+            moveToBlockedLocationAfterReactStateUpdate(location);
+          },
+        });
+      },
+
+      onDiscard: () => {
+        dashboard.exitEditMode({ skipConfirm: true });
+        hideModal();
+        moveToBlockedLocationAfterReactStateUpdate(location);
+      },
+      onDismiss: hideModal,
+    });
+
+    return false;
+  };
+
+  return <Prompt when={true} message={onHistoryBlock} />;
+});
+
+DashboardPrompt.displayName = 'DashboardPrompt';
+
+function moveToBlockedLocationAfterReactStateUpdate(location?: H.Location | null) {
+  if (location) {
+    setTimeout(() => locationService.push(location), 10);
+  }
+}
+interface UnsavedChangesModalProps {
+  onDiscard: () => void;
+  onDismiss: () => void;
+  onSaveDashboardClick?: () => void;
+}
+
+export const UnsavedChangesModal = ({ onDiscard, onDismiss, onSaveDashboardClick }: UnsavedChangesModalProps) => {
+  const styles = getStyles();
+  return (
+    <Modal
+      isOpen={true}
+      title="Unsaved changes"
+      onDismiss={onDismiss}
+      icon="exclamation-triangle"
+      className={styles.modal}
+    >
+      <h5>Do you want to save your changes?</h5>
+      <Modal.ButtonRow>
+        <Button variant="secondary" onClick={onDismiss} fill="outline">
+          Cancel
+        </Button>
+        <Button variant="destructive" onClick={onDiscard}>
+          Discard
+        </Button>
+        <Button onClick={onSaveDashboardClick}>Save dashboard</Button>
+      </Modal.ButtonRow>
+    </Modal>
+  );
+};
+
+const getStyles = () => ({
+  modal: css({
+    width: '500px',
+  }),
+});
+
+/**
+ * For some dashboards and users changes should be ignored *
+ */
+export function ignoreChanges(current: DashboardScene | null, original?: Dashboard) {
+  if (!original) {
+    return true;
+  }
+
+  // Ignore changes if original is unsaved
+  if ((original as Dashboard).version === 0) {
+    return true;
+  }
+
+  // Ignore changes if the user has been signed out
+  if (!contextSrv.isSignedIn) {
+    return true;
+  }
+
+  if (!current) {
+    return true;
+  }
+
+  const { canSave, fromScript, fromFile } = current.state.meta;
```

Comment:
❓ Out of curiosity, what is the meaning of these "fromScript" and "fromFile" props?  🙈